collector/kernel/kernel_collector_restarter.h

```diff
@@ -0,0 +1,46 @@
+//
+// Copyright 2021 Splunk Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#pragma once
+
+#include <platform/platform.h>
+
+class KernelCollector;
+
+/**
+ * Helper class to facilitate restarting KernelCollector, i.e. when BufferedPoller detects lost BPF samples (PERF_RECORD_LOST).
+ * If KernelCollector startup has completed, then a restart request will be processed immediately.
+ * If KernelCollector startup has not completed, then the processing of a restart request will be deferred until after
+ * KernelCollector startup has completed.
+ * If a restart is already in progress, subsequent restart requests will be ignored.
+ */
+class KernelCollectorRestarter {
+  friend class KernelCollector;
+
+public:
+  KernelCollectorRestarter(KernelCollector &collector);
+  void startup_completed();
+  void request_restart();
+
+private:
+  void check_restart();
+
+  bool startup_completed_;
```

Comment:
Are these states able to overlap?  Or is this more of a state machine sort of behavior?  If the later, might an enum be preferable?

Reply:
The states can overlap so it is not a simple state machine.  IOW, it is possible for a restart request to come in before or after startup is completed, but we don't want to restart while still starting up, so the restart will occur once both startup_completed_ and restart_requested_ are true..  (It's probably unlikely for a restart request to come in due to lost BPF samples before startup has completed, but it is possible, especially in test environments.)

Reply:
Thank you for the clarification!